### PR TITLE
ARROW-6069: [Rust] [Parquet] Add converter.

### PR DESCRIPTION
**Pull request by liurenjie1024**
Converter is used convert record reader's content to arrow primitive array.

### CODE REVIEW

**Comment by liurenjie1024** [comment]
@sunchao @nevi-me @andygrove  Please take a look when you get a chance.

File: rust/parquet/src/arrow/converter.rs [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+convert_as!(i32, i8);
+convert_as!(i32, i16);
+convert_as!(i32, u8);
+convert_as!(i32, u16);
+convert_as!(i32, u32);
+
+/// Builder converter is used when parquet's data type is different from
+/// arrow's native type. In this case, we need to iterate over values in record reader
+/// and convert parquet value into arrow value one by one.
```
Comment (nevi-me) [inline rust/parquet/src/arrow/converter.rs]: Is the performance different if say you build an `Int32Array`, and then use the `cast` kernels to convert to an appropriate type? If a cast works out faster, then we could drop the `Converter` and remain with `DirectConverter`.

File: rust/parquet/src/util/test_common/page_util.rs [hunk 2]
```diff
@@ -155,3 +157,21 @@ impl DataPageBuilder for DataPageBuilderImpl {
         }
     }
 }
+
+pub struct InMemoryPageReader {
```
Comment (nevi-me) [inline rust/parquet/src/util/test_common/page_util.rs]: Do we need to document this test page reader?

**Review by nevi-me — COMMENTED** [review]
Hi @liurenjie1024, I've had a chance to look at your PR, have a few minor comments/questions.

File: rust/parquet/src/arrow/converter.rs (hunk 1 above)
Comment (liurenjie1024) [reply rust/parquet/src/arrow/converter.rs]: Good point! Will try to use cast kernel.

File: rust/parquet/src/arrow/converter.rs [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+use arrow::array::PrimitiveArray;
+use std::marker::PhantomData;
+use std::mem::transmute;
+
+use crate::data_type::BoolType;
+use crate::data_type::DoubleType as ParquetDoubleType;
+use crate::data_type::FloatType as ParquetFloatType;
+use crate::data_type::Int32Type as ParquetInt32Type;
+use crate::data_type::Int64Type as ParquetInt64Type;
+use arrow::datatypes::BooleanType;
```
[the remark this answers is not in the document]
Comment (liurenjie1024) [reply rust/parquet/src/arrow/converter.rs]: Thanks for pointing this out. But I used `cargo fmt` and it doesn't change anything. So I'm confusing how to consolidate these? Or should we just leave these formatting things to `cargo fmt`?

File: rust/parquet/src/util/test_common/page_util.rs (hunk 2 above)
Comment (liurenjie1024) [reply rust/parquet/src/util/test_common/page_util.rs]: Fixed.

File: rust/parquet/src/arrow/converter.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (sunchao) [reply rust/parquet/src/arrow/converter.rs]: You can replace these imports with star import to save some space, although I'm not sure what's the recommended style for this (in Java usually it's not recommended to have star imports).

[the remark this answers is not in the document]
Comment (nevi-me) [reply rust/parquet/src/arrow/converter.rs]: cargo fmt won't have an issue, but it's better to keep imports nested as it reduces LOC, and would consistent with the rest of the code base

File: rust/parquet/src/arrow/converter.rs [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+}
+
+impl<ParquetType, ArrowSourceType, ArrowTargetType> Converter<ParquetType>
+    for CastConverter<ParquetType, ArrowSourceType, ArrowTargetType>
+where
+    ParquetType: DataType,
+    ArrowSourceType: ArrowPrimitiveType,
+    ArrowTargetType: ArrowPrimitiveType,
+{
+    fn convert(record_reader: &mut RecordReader<ParquetType>) -> Result<Arc<Array>> {
```
Comment (sunchao) [inline rust/parquet/src/arrow/converter.rs]: Was looking at how we can avoid iterating this value by value - seems in C++ it just does a `reinterpret_cast` on the values decoded from Parquet, which is interesting: https://github.com/apache/arrow/blob/master/cpp/src/parquet/arrow/reader_internal.cc#L575

Comment (liurenjie1024) [reply rust/parquet/src/arrow/converter.rs]: This is used when arrow's data type is same as parquet's physical type. In cast kernel's implements, the one by one conversion only happens when ArrowSourceType is different from ArrowTargetType.

**Comment by liurenjie1024** [comment]
@nevi-me @sunchao Fixed comments, please take a look.

File: rust/parquet/src/arrow/converter.rs [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+use arrow::datatypes::Int8Type;
+use arrow::datatypes::UInt16Type;
+use arrow::datatypes::UInt32Type;
+use arrow::datatypes::UInt64Type;
+use arrow::datatypes::UInt8Type;
+
+/// A converter is used to consume record reader's content and convert it to arrow
+/// primitive array.
+pub trait Converter<T: DataType> {
+    /// This method converts record reader's buffered content into arrow array.
```
[the remark this answers is not in the document]
Comment (liurenjie1024) [reply rust/parquet/src/arrow/converter.rs]: I added some comments about post condition. As with precondition, in fact current implementation doesn't require anything special. Even if record reader is not filled, the result is an empty array.

File: rust/parquet/src/arrow/converter.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (liurenjie1024) [reply rust/parquet/src/arrow/converter.rs]: Fixed.

File: rust/parquet/src/arrow/converter.rs (hunk 4 above)
Comment (sunchao) [reply rust/parquet/src/arrow/converter.rs]: No. This is also used when Arrow type is narrower than Parquet type, for instance, converting Parquet's Int32 to Arrow's UInt8. See https://github.com/apache/arrow/blob/master/cpp/src/parquet/arrow/reader_internal.cc#L1043.

I think what you have is a good start - we can think about ways to optimize this later.

**Review by sunchao — APPROVED** [review]

File: rust/parquet/src/arrow/converter.rs (hunk 4 above)
Comment (liurenjie1024) [reply rust/parquet/src/arrow/converter.rs]: I think the magic hides in std::copy

**Review by nevi-me — APPROVED** [review]

**Pull request closed without merging**